DeepSeek on AWS (#2632)
huggingface/blog: deepseek-aws.md

Let’s review how you can deploy and fine-tune DeepSeek R1 models with Hugging Face on AWS.
- [Deploy DeepSeek R1 models](#deploy-deepseek-r1-models)
    - [Deploy on Hugging Face Inference Endpoints on AWS](#deploy-on-hugging-face-inference-endpoints-on-aws)
    - [Deploy on Amazon Sagemaker AI with Hugging Face LLM DLCs](#deploy-on-amazon-sagemaker-ai-with-hugging-face-llm-dlcs)
        - [DeepSeek R1 on GPUs](#deepseek-r1-on-gpus)
        - [Distilled models on GPUs](#distilled-models-on-gpus)
        - [Distilled models on Neuron](#distilled-models-on-neuron)
    - [Deploy on EC2 Neuron with the Hugging Face Neuron Deep Learning AMI](#deploy-on-ec2-neuron-with-the-hugging-face-neuron-deep-learning-ami)
- [Fine-tune DeepSeek R1 models](#fine-tuning-deepseek-r1-models)
    - [Fine tune on Amazon Sagemaker AI with Hugging Face Training DLCs](#fine-tuning-on-amazon-sagemaker-ai-with-hugging-face-training-dlcs)
    - [Fine tune on EC2  Neuron with the Hugging Face Neuron Deep Learning AMI](#fine-tuning-on-ec2--neuron-with-the-hugging-face-neuron-deep-learning-ami)

## Deploy DeepSeek R1 models

### Deploy on Hugging Face Inference Endpoints on AWS

> [**Hugging Face Inference Endpoints**](https://ui.endpoints.huggingface.co/) offers an easy and secure way to deploy Machine Learning models for use in production on AWS. Inference Endpoints empower developers and data scientists alike to create AI applications without managing infrastructure: simplifying the deployment process to a few clicks, including handling large volumes of requests with autoscaling, reducing infrastructure costs with scale-to-zero, and offering advanced security.

Reviewer: ```suggestion
[**Hugging Face Inference Endpoints**](https://ui.endpoints.huggingface.co/) offers an easy and secure way to deploy Machine Learning models on dedicated compute for use in production on AWS. Inference Endpoints empower developers and data scientists alike to create AI applications without managing infrastructure: simplifying the deployment process to a few clicks, including handling large volumes of requests with autoscaling, reducing infrastructure costs with scale-to-zero, and offering advanced security.
```

With Inference Endpoints, you can deploy any of the 6 distilled models from DeepSeek-R1 and also a quantized version of DeepSeek R1 made by Unsloth: https://huggingface.co/unsloth/DeepSeek-R1-GGUF.

> On the model page, click on Deploy, then on HF Inference Endpoints. You will be redirect to the Inference Endpoint page, where we selected for you an optimized inference container, and the recommended hardware to run the model. Once you created your endpoint, you can send your queries to DeepSeek R1 for 8.3$ per hour with AWS 🤯.

Reviewer: ```suggestion
On the model page, click on Deploy, then on HF Inference Endpoints. You will be redirected to the Inference Endpoint page, where we selected for you an optimized inference container, and the recommended hardware to run the model. Once you created your endpoint, you can send your queries to DeepSeek R1 for 8.3$ per hour with AWS 🤯.

You can find DeepSeek R1 and distilled models, as well as other popular open LLMs, ready to deploy on optimized configurations in the [Inference Endpoints Model Catalog](https://endpoints.huggingface.co/catalog?task=text-generation).
```

```python
model_id = 'deepseek-ai/DeepSeek-R1-Distill-Llama-70B'
model_name = hf_model_id.split("/")[-1].lower()

# Hub Model configuration. https://huggingface.co/models
hub = {
	'HF_MODEL_ID': model_id,
	'SM_NUM_GPUS': json.dumps(8)
}

# create Hugging Face Model Class
huggingface_model = HuggingFaceModel(
	image_uri=get_huggingface_llm_image_uri("huggingface",version="3.0.1"),
	env=hub,
	role=role, 
)
```

Deploy the model to a Sagemaker endpoint and test the endpoint:

```python
endpoint_name = f"{model_name}-ep"

# deploy model to SageMaker Inference
predictor = huggingface_model.deploy(
  endpoint_name=endpoint_name,
	initial_instance_count=1,
	instance_type="ml.g6.48xlarge",
	container_startup_health_check_timeout=2400,
  )
  
# send request
predictor.predict({"inputs": "What is the meaning of life?"})
```

That’s it, you deployed a Llama 70B reasoning model! 

Because you are using a TGI v3 container under the hood, the most performant parameters for the given hardware will be automatically selected.

Make sure you delete the endpoint once you finished testing it.

```python
predictor.delete_model()
predictor.delete_endpoint()
```

#### Distilled models on Neuron

> Let’s walk through the deployment of DeepSeek-R1-Distill-Llama-70B on a Neuron instance.

Reviewer: ```suggestion
Let’s walk through the deployment of DeepSeek-R1-Distill-Llama-70B on a Neuron instance, like AWS Trainium 2 and AWS Inferentia 2.
```

```bash
docker run -p 8080:80 \
       -v $(pwd)/data:/data \
       --device=/dev/neuron0 \
       --device=/dev/neuron1 \
       --device=/dev/neuron2 \
       --device=/dev/neuron3 \
       --device=/dev/neuron4 \
       --device=/dev/neuron5 \
       --device=/dev/neuron6 \
       --device=/dev/neuron7 \
       --device=/dev/neuron8 \
       --device=/dev/neuron9 \
       --device=/dev/neuron10 \
       --device=/dev/neuron11 \
       -e HF_BATCH_SIZE=4 \
       -e HF_SEQUENCE_LENGTH=4096 \
       -e HF_AUTO_CAST_TYPE="bf16" \
       -e HF_NUM_CORES=24 \
       ghcr.io/huggingface/neuronx-tgi:latest \
       --model-id deepseek-ai/DeepSeek-R1-Distill-Llama-70B \
       --max-batch-size 4 \
       --max-total-tokens 4096
```

It will take a few minutes to download the compiled model from the Hugging Face cache and launch a TGI endpoint.

Then, you can test the endpoint: 

```bash
       curl localhost:8080/generate \
       -X POST \
       -d '{"inputs":"Why is the sky dark at night?"}' \
       -H 'Content-Type: application/json'
```

Make sure you pause the EC2 instance once you are done testing it.

> | **Note:** The team is working on enabling all DeepSeek models deployment on Trainium & Inferentia with the Hugging Face Neuron Deep Learning AMI. Stay tuned! 

Reviewer 2: If you mean the distilled ones, then it is already available.

Author: I specified deepseek r1
